Maximum Bucket reducer

colings86 opened the pull request:
This reducer calculates the bucket(s) in a sibling aggregation which contains the maximum value of a particular metric. The request is of the form:

```javascript
GET test/_search?search_type=count
{
  "aggs": {
    "my_date_histo": {
      "date_histogram": {
        "field": "d",
        "interval": "day",
        "min_doc_count": 0
      },
      "aggs": {
        "avg_i": {
          "avg": {
            "field": "i"
          }
        }
      }
    },
    "max_avg_i": {
      "max_bucket": {
        "bucketsPath": "my_date_histo>avg_i"
      }
    }
  }
}
```

Which will produce the following response:

```javascript
{
   "took": 256,
   "timed_out": false,
   "_shards": {
      "total": 5,
      "successful": 5,
      "failed": 0
   },
   "hits": {
      "total": 10,
      "max_score": 0,
      "hits": []
   },
   "aggregations": {
      "my_date_histo": {
         "buckets": [
            {
               "key_as_string": "2014-01-01T00:00:00.000Z",
               "key": 1388534400000,
               "doc_count": 1,
               "avg_i": {
                  "value": 1,
                  "value_as_string": "1.0"
               }
            },
            ...
            {
               "key_as_string": "2014-01-10T00:00:00.000Z",
               "key": 1389312000000,
               "doc_count": 1,
               "avg_i": {
                  "value": 10,
                  "value_as_string": "10.0"
               }
            }
         ]
      },
      "max_avg_i": {
         "value": 10,
         "keys": [
            "2014-01-10T00:00:00.000Z"
         ]
      }
   }
}
```

jpountz commented on src/main/java/org/elasticsearch/search/query/QuerySearchResult.java:
```diff
@@ -114,6 +120,14 @@ public void aggregations(InternalAggregations aggregations) {
         this.aggregations = aggregations;
     }
 
+    public List<SiblingReducer> reducers() {
```
Can it really only work for siblings or is it just that we only have siblings for now?

colings86 replied:
I think it can only work for siblings since this is only concerned with top-level reducers and they must be sibling reducers since they have no parent?

polyfractal commented on src/main/java/org/elasticsearch/search/aggregations/reducers/bucketmetrics/MaxBucketReducer.java:
```diff
@@ -0,0 +1,163 @@
+        return new InternalBucketMetricValue(name(), keys, maxValue, formatter, Collections.EMPTY_LIST, metaData());
+    }
+
+    private Double resolveBucketValue(InternalMultiBucketAggregation multiBucketAggregation, InternalMultiBucketAggregation.Bucket bucket,
```
Any way we can unify this with `BucketHelpers.resolveBucketValue()`?    Or perhaps move this into the helper class and rename both of them to be more specific (`resolveHistoBucketValue()` and `resolveMultiBucketValue()` or something?)

Also, I foresee this needing to handle gap policies too...unfortunately. :(  For example, an agg like Autocorrelation needs to ingest a histogram (which might need gap policy) but emits a set of sibling buckets that represent correlation lags.

polyfractal replied:
Ignore the last bit about gap policies...I was confused about how this bit of code worked.  Started playing with it and cleared up my confusion :)

colings86 replied:
I actually thought I already had unified this with the method in BucketHelpers but will look. Thanks for catching

polyfractal commented on src/main/java/org/elasticsearch/search/aggregations/reducers/bucketmetrics/MaxBucketReducer.java:
```diff
@@ -0,0 +1,163 @@
+            } else if (propertyValue instanceof InternalNumericMetricsAggregation.SingleValue) {
+                return ((InternalNumericMetricsAggregation.SingleValue) propertyValue).value();
+            } else {
+                throw new AggregationExecutionException(DerivativeParser.BUCKETS_PATH.getPreferredName()
```
Small typo...I think this was supposed to be  `MaxBucketParser.BUCKETS_PATH` ?

colings86 replied:
yep, good catch

polyfractal commented:
Just two general things I ran into.

First, when I started to implement tests for the ACF agg, I discovered that `ReducerBuilder` needs to extend `AbstractAggregationBuilder`, or else it cannot be used as a top-level aggregation in Java queries.  E.g. this code is not possible without extending:

```java
SearchResponse response = client()
                .prepareSearch("idx").setTypes("linear")
                .addAggregation(histogram("histo").field(INTERVAL_FIELD).interval(interval))        
                .addAggregation(acf("acf").setBucketsPaths("histo>the_max"))  // <-- this won't compile
                .execute().actionGet();
```

In my branch I extended `ReducerBuilder` and the change seemed fairly minor, and does not appear to break things.  But I'm not sure if it's a good idea or not, if things should be structured differently, etc.

The second issue was also related to SiblingReducers being potentially "top level aggregations".  I had to implement an `InternalACF` class so that agg responses could be serialized back to the client, as well as register the stream as an "aggregation".  So the agg has two registered streams (reducer + "aggregation").

Also doesn't seem like a big deal, but thought I should raise the issue.

colings86 commented:
@polyfractal yep the first issue is my fault, I forgot to update the search builder to accept ReducerBuilders, I'll update that in my Sibling reducers PR.

You are correct with the second issue, you will need to register two streams (reducer and aggregation). this is because the reducer has to be serialised to get it back from the shards (where the parsing of the request is currently done) for the reduce phase to have it and the aggregation has to be serialised because this is the response that is needed for the Java API. This should be helped once https://github.com/elastic/elasticsearch/issues/10217 has been merged as we should hopefully be able to get rid of the serialising of the reducers, but we will have to see once it's merged in.

uboness commented on src/main/java/org/elasticsearch/search/aggregations/reducers/Reducer.java:
```diff
@@ -66,6 +69,13 @@
 
     }
 
+    public static final Function<Aggregation, InternalAggregation> FUNCTION = new Function<Aggregation, InternalAggregation>() {
```
not in this PR, but in this file... can you change `ParseField("bucketsPath")` to `ParseField("buckets_path")`

please don't use camelCasing

colings86 replied:
Yep, sorry, still trying to get out of the habit of camelCasing :)

jpountz replied:
Can you rename FUNCTION to something more explicit?

polyfractal commented:
@colings86 great, thanks for clarifying those issues.  Other than that, the new sibling stuff worked great, was very happy how smooth implementing the ACF went.

colings86 commented:
@polyfractal I have pushed a commit which should address your comments

colings86 commented:
@jpountz would you mind re-reviewing this when you have the time? I have rebased the commits from the feature/aggs_2_0 branch since it was getting quite out of date. I also added a simple test. It will obviously need more tests before it is considered feature complete but I am keen to get this merged into the feature/aggs_2_0 branch ASAP.

jpountz commented on src/main/java/org/elasticsearch/action/percolate/PercolateShardResponse.java:
```diff
@@ -69,6 +72,9 @@ public PercolateShardResponse(BytesRef[] matches, List<Map<String, HighlightFiel
             }
+            if (result.reducers() != null) {
+                this.reducers = result.reducers();
+            }
```
The `null` check does not look useful, is it?

jpountz commented on src/main/java/org/elasticsearch/action/percolate/PercolateShardResponse.java:
```diff
@@ -112,6 +118,10 @@ public InternalAggregations aggregations() {
 
+    public List<SiblingReducer> reducers() {
+        return reducers;
+    }
```
I think these reducers should be (de)serialized in readFrom/writeTo? Maybe it's worth having a percolation with reducers test to make sure everything works fine.

colings86 replied:
Added a percolator test and implemented read/write methods correctly to serialise the reducers

jpountz commented on src/main/java/org/elasticsearch/percolator/PercolatorService.java:
```diff
@@ -826,15 +847,29 @@ private InternalAggregations reduceAggregations(List<PercolateShardResponse> sha
+        if (aggregations != null) {
+            List<SiblingReducer> reducers = shardResults.get(0).reducers();
+            if (reducers != null) {
+                List<InternalAggregation> newAggs = new ArrayList<>(Lists.transform(aggregations.asList(), Reducer.FUNCTION));
```
Side note: we need to rename Reducer.FUNCTION to have a more explicit name.

colings86 replied:
Agreed, although ideally I would like to get rid of it since all we are doing is using it to cast a list of one type to a list of another type. I just haven't worked out how I can get rid of it yet. Any thoughts on that are welcome :)

jpountz commented on src/main/java/org/elasticsearch/search/aggregations/reducers/bucketmetrics/MaxBucketReducer.java:
```diff
@@ -0,0 +1,144 @@
+
+    public InternalAggregation doReduce(Aggregations aggregations, ReduceContext context) {
+        List<String> maxBucketKeys = new ArrayList<>();
+        double maxValue = Double.MIN_VALUE;
```
s/MIN_VALUE/NEGATIVE_INFINITY/ ?

jpountz commented on src/main/java/org/elasticsearch/search/aggregations/reducers/bucketmetrics/MaxBucketReducer.java:
```diff
@@ -0,0 +1,144 @@
+    public InternalAggregation doReduce(Aggregations aggregations, ReduceContext context) {
+        List<String> maxBucketKeys = new ArrayList<>();
+        double maxValue = Double.MIN_VALUE;
+        List<String> bucketsPath = AggregationPath.parse(bucketsPaths()[0]).getPathElementsAsStringList();
```
Only `bucketsPaths()[0]` is used. Should we validate that only path has been provided? (maybe it exists and I just don't see it?)

jpountz replied:
just saw it in the factory validation, nevermind :)

jpountz commented on src/main/java/org/elasticsearch/search/aggregations/reducers/bucketmetrics/MaxBucketReducer.java:
```diff
@@ -0,0 +1,144 @@
+                            maxBucketKeys.clear();
+                            maxBucketKeys.add(bucket.getKeyAsString());
+                            maxValue = bucketValue;
+                        } else if (bucketValue == maxValue) {
```
You need to use .equals on a Double

jpountz commented on src/main/java/org/elasticsearch/search/controller/SearchPhaseController.java:
```diff
@@ -391,6 +407,19 @@ public InternalSearchResponse merge(ScoreDoc[] sortedDocs, AtomicArray<? extends
         }
 
+        if (aggregations != null) {
+            List<SiblingReducer> reducers = queryResults.get(0).value.queryResult().reducers();
```
Let's reuse the `firstResults` variable instead?

jpountz commented:
Sure @colings86 I left some comments.

I think we should merge the AggregationBuilder and ReducerBuilder classes. Although we deal with them differently on server-side, they are the same from a client perspective? (but we could still keep the abstract impls that we have in order not to have to duplicate the subAggregation or bucketsPath methods everywhere)

Also this PR made me realize the ReduceContext API is not very friendly as sometimes you need to pass `null` to the wrapped aggregations. So maybe we should remove aggregations from ReduceContext and change InternalAggregation.reduce to take a list of aggregations in addition to the ReduceContext. (can be done in a different PR, just thinking out loud)

colings86 commented:
@jpountz are you talking about making ReducerBuilder extend AggregationBuilder. The problem I see with that is that Aggregations are allowed sub-aggregations but reducers are not. The advantage with having the ReducerBuilder separate is that it is not possible in the Java API to create a reducer with sub-aggregations/reducers. I suppose we could override those methods in ReducerBuilder and have them throw an UnsupportedOperationExeception or leave it to the parser to detect the illegal behaviour?

I was actually going to make that change to the ReduceContext already as I saw the same thing. Even before we introduce the reducers framework there are many instances in the code where we pass in null for the aggregations when constructing the ReduceContext because we don't need it at that point. I will do that in a separate PR after this is merged (if that works for you?).

jpountz commented:
> @jpountz are you talking about making ReducerBuilder extend AggregationBuilder.

Oh, I just noticed that we already have a base class: AbstractAggregationBuilder. So maybe we just need to make ReducerBuilder extend it and then we don't need to add a new method to eg. PercolateRequestBuilder?

> I was actually going to make that change to the ReduceContext already as I saw the same thing. Even before we introduce the reducers framework there are many instances in the code where we pass in null for the aggregations when constructing the ReduceContext because we don't need it at that point. I will do that in a separate PR after this is merged (if that works for you?).

Awesome. :)

colings86 commented:
@jpountz I pushed a new commit with the builder changes we discussed

jpountz commented:
LGTM
